### DISCUSSION TITLE
What is even worth doing?

### DISCUSSION THREAD

**Discussion by MatthewCaseres** [actuarialopensource/website#42] [General]
Trying to think through some work items that are interesting, a lot of work can be done but it isn't clear what should be done. In no order.

### Lifelib

* make a reasonable contribution to lifelib
    * Basic structure of actuarial programs, introduction to caching, graph capture.
    * https://github.com/benfred/py-spy should allow for some analysis of seriatim vs. array based methods.
        * The issue is probably related to just function call overall. We save so many recursive calls with the array, how to communicate this.
    * finish up analysis of kahn's algorithm for optimizing memory. I don't think trying to publish a library or something like that is necessary, just a solid analysis.
    * Using autodiff with a decorator.
    
Maybe even make a PDF document. Keep it short and rigorous.
    
So the pitch is to contribute a couple of high quality educational notebooks to the lifelib projects so that people know some basic information about building their own simple modeling systems in Python.

### Website

The fact that star counts have not been updating is pretty bad. We need to nuke the source code and make it as concise and maintainable as possible.

### Simulating universal life data

This also has potential as a lifelib contribution. We need the simulated data to do reproducible research.

### Benchmarks

The benchmarks are fine but I don't think there are any enhancements that need to be prioritized. 

### Chainladder

Raising issues and not fixing is in poor taste. Should probably come back around and solve the issues I raised.

**Comment by lewisfogden** [comment]
I'm working on an approach to vectorize access to multi-dimensional tables (e.g. with keys of age, duration, sex) - I could put together a problem statement for this and some examples?

It's key blocker for me in terms of using Python in production (most of our tables have 2 or 3 keys, included banded keys).

(non-vertorized tables work fine using pandas MultiIndexes, but vectorised Indexes only work correctly in pandas for a single index)

**Reply by MatthewCaseres** [reply]
example of an age,duration,mortality table index here - https://github.com/actuarialopensource/benchmarks/blob/main/Python/mortality.py#L24. Perhaps we can do better. My current approach would basically be to use array indexing as shown in the link, and then test the heck out of it in PyTest to cover potential edge cases.

Scenarios that involve more complicated assumption structures will be really valuable, please send any thoughts you have.

**Reply by lewisfogden** [reply]
I wrote up a bit of a spec (in examples folder in below repo) but then ended up coding up a solution.

I've used column suffixes to capture metadata.  Haven't coded up the bounded/clipped version yet (which is nice for select/ultimate mortality tables)

https://github.com/lewisfogden/heavytables

**Reply by MatthewCaseres** [reply]
Very helpful. I'll give it a go this afternoon.

**Reply by MatthewCaseres** [reply]
I see two different things

**Doing joins on a range/band**: I see you have an approach for this, but it might not generalize for like a `band_band_band` column with arbitrary bandings? My intuition is that we want an inequality join. Pandas does not have a native way to do this. Dplyr in R does with the newish `join_by` https://dplyr.tidyverse.org/reference/join_by.html. In Polars they are currently tracking it in this issue https://github.com/pola-rs/polars/issues/10068. It is implemented in DuckDB which should interop with dataframes nicely, https://duckdb.org/2022/05/27/iejoin.html. I had not thought about this detail before, thank you.

**Storage formats for assumption tables**: You want to use Excel sheets to store assumptions. This seems reasonable I don't really have any complaints about that. I don't know that the type inference when loading excel is so bad that we need to suffix each column with the type?

My concern is that we are going to end up with a framework that attempts to do something that is somewhat standard in Python like joining a ModelPoints file with an Assumption table. I don't think a library needs to be built, I see the appeal to a standard format for assumption tables, but there might be a sacrifice in flexibility. For example, would you be able to load a whole set of mortality tables select and ultimate and perform a calculation using them? And then the framework will have a bunch of custom code for this one scenario. And the framework grows and grows to handle each scenario you encounter but in each case maybe a few lines of code with no framework would be fine? I think there needs to be a better body of knowledge surrounding methods for performing the tasks you are describing. I'm recently a little frustrated with how we have so few models and so many frameworks. 

I'll start working on a notebook on assumption tables. Homework assignments + work could delay delivery, but it should not take more than 2 weeks.

edit: If you have a lot of assumption tables I see the appeal of having a standard format for representing them in Excel.

**Reply by MatthewCaseres** [reply]
sorry for not replying with code, but this is basically my thoughts. I will restrain myself from writing any more prose here until I have an open pull request in lifelib.

## Storing assumption tables in excel

Yes I like this idea. Actuaries use excel, they will enjoy this. Including the type information probably not a bad thing.

Both the `clipping` and `bands` are really sort of the same, because clipping can be expressed as banding I believe. Not sure I would personally use a single `band` column and then make assumptions about the sortedness of it. It introduces complexity that can be avoided by doing something more ugly and having two columns. But using the two columns is actually a very general solution so why not.

<img width="484" alt="Screenshot 2024-03-13 at 9 53 41 PM" src="https://github.com/actuarialopensource/actuarialopensource.org/assets/43053796/414e0cd4-b11c-4f21-8881-d4af9a50c2bb">

And then use the inequality join.

## indexing into arrays

You don't need to index into the values during the execution of the model. You can take the modelpoint file, join the necessary quantities to it, and then access this column of the "enriched modelpoints" whenever you want. I do not think any indexing needs to be done for assumptions that are not time dependent.

You will might end up indexing into a table during model execution if your assumption is dependent on time. shorthand, this is a TD (time dependent) assumption. There are **TD axes and non TD axes of the TD assumption**. The non TD axes can be combined into a single linear index or not, but it probably cleans up the indexing code to make the linear index, I would make the linear index. Then there is the time dependent assumption.

TD assumptions then can be indexed into like `arr[linear_index_array, duration]` or something like that, but this arr is just a numpy array.

## Trying to build tools for actuaries

This is just an aside and maybe not one I am right about. Actuaries are probably great at using their proprietary software and generally not that good at Python. So I get the desire to make them feel at home in Python by creating feature parity. 

But I don't know that creating feature parity with proprietary software is Pythonic. Because then we are implementing our own code navigation systems, debuggers, dataframe objects, or asking people to structure the files in their project in a certain way.

And proprietary modeling system formulas rely on some implicit global state anyways, not sure how to feel about that. `pols_if(t)` but clearly isn't parameterized by only `t`. But then if we don't bless a certain way inject global state related to modelpoints into our functions it isn't considered idiomatic to actuaries. But that is a bad idea!

Trying to build tools for actuaries because it decreases the quality of the modeling system (not Pythonic), in some cases impacts the testability (they want global state), and increases the maintenance cost (they want features they don't need to calculate the numbers they need to calculate). And then, if we give them what they wanted for free, they won't be able to maintain it because our codebase is written in good Python and all they know is the dumbed down stuff we gave them. It isn't a sustainable model of development.

**Reply by lewisfogden** [reply]
So will give you a bit of my experience (background: I lead a model development team for a large UK life insurer which employs around 300 actuaries, and my team are responsible for developing and maintaining 4 or so material models as well as lots of lower material tools & models, we use a range of proprietary packages as well as Python & R).

I have built two complete models using my framework (heavylight) which we've used for testing and validation of other models, one of these has transitioned to be come the main model and been subject to independent validation / audit etc, so it is possible to achieve production use.   (can't share source unfortunately due to intellectual property rights)

However, Python is still very much a scary new thing for most actuaries, and the idea of open source modelling is treated with suspicion (being risk averse and all), and although most actuaries are coding one way or another (excel, VBA, proprietary models) there is still a big gap in confidence and skill to develop using a programming language.  Getting people to install Python and run the script to execute the model can be a significant barrier.

So given that, there is a need to provide a bit more of a guided approach - I think before we can do that we need to get the basics right in terms of use cases and requirements, and demonstrate that FOSS is as good as proprietary.

Key for me are:
- a consistent framework: clear specifications on how models should be written (to avoid spaghetti code)
- readability of code to understand functionality (hence why looking at Python rather than C++ or Rust)
- change control: can it be managed in git etc
- ability to debug: can I see code alongside outputs (like diagram view in some tools)
- ability to deal with various tables (e.g. CMI mortality & morbidity, expenses, yield curves)
- ability for users to run easily for different scenarios and configure run details etc
- performance (pure python is slow - and heavylight is pure python, but vectorising is one solution here, cython is another)
- interaction with inputs, e.g. Excel, csvs
- Licence (e.g. MIT) and IT security risk management considerations
- Total cost of ownership - while 'free' to acquire/install, developer time is the main limitation.

I've seen open source models developed without considering these, and the outcome has been negative (I spent 3 weeks trying to debug one model).  Main issue is people using R/Python and thinking that dataframes are the solution to everything, whereas dataframes are for storing and analysing columnar data, which is generally an output from an actuarial model.

**Reply by lewisfogden** [reply]
I've put together a working model (with lots of random features!) to show my approach:

https://github.com/lewisfogden/heavytables/blob/main/examples/heavylight_test.py

(Note that the 'fake' table is based on the structure of the IFoA CMI tables, I've filled it with fake values as the CMI data is copyrighted.)

Running with 100k policies is very quick - couple of plots of the results. 
![image](https://github.com/actuarialopensource/actuarialopensource.org/assets/8605587/269f5f4e-d8fd-46b4-90b5-aa87d2fe57b6)

![image](https://github.com/actuarialopensource/actuarialopensource.org/assets/8605587/928254fc-fcc2-4c01-bff5-ff92ecfeab35)

(I need to do some code to summarise the outputs properly, e.g. sum up each variable over the policies, or extract single policies)

**Reply by MatthewCaseres** [reply]
My negativity from earlier (on various topics) was totally unjustified. I feel bad about that! But, oh well.

This is great work, I hadn't considered things like looking up a band at runtime with the np search sorted. So ironic I try to explain linearizing an index or something for time dependent assumptions without understanding what you were doing, and then you had actually linearized the index further than I had considered.

I'll try to contribute something to the tables repo, set up some automated tests.

Could you help me understand why you don't use a decorator to cache? I guess, what is the barrier to adoption there compared to your current class?

**Reply by MatthewCaseres** [reply]
Mostly I'm writing some notebooks about how to design decorators and if there is something seriously wrong with them I would be wasting my time at the moment.

**Reply by lewisfogden** [reply]
No problem at all - I didn't think you were negative, just a good bit of challenge 🙂

So, I do use decorators!, but I've hidden them in the class initialisation, when the class initialises each method is scanned, and the Cache decorator applied.

<img width="791" alt="Screenshot 2024-03-15 at 21 49 25" src="https://github.com/actuarialopensource/actuarialopensource.org/assets/8605587/966f45c2-4f2e-401e-99e2-e1004e46c75e">

My rationale for using classes is that it keeps all the model logic together, and you don't need to worry about emptying caches etc - I original wrote the model as functions with a decorator for each, but I didn't like it.  My original framework (heavymodel) is really designed about modelling a single policy at a time, and comparing policies and models by having multiple instances.

https://github.com/lewisfogden/heavylight/blob/main/src/heavylight/heavylight.py (Source)

**Reply by MatthewCaseres** [reply]
Downsides of using classes:
* Limited ability to organize formulas within different files? Concern with scaling complexity of model
Ways to get around downside of using classes:
* Unsure (any thoughts?)

I see the downsides of decorators, I guess I'm not convinced that the scales tip in favor of a class based approach. The decorator based approach does suffer from the dependency on global state. I'll be thinking about this.

**Reply by lewisfogden** [reply]
Main downside (which isn't too bad) is having to refer to methods and data with `self.` at the start, which reduces the readability a bit (and might confuse beginners a bit)

Complexity of model - my view is if you go beyond about 200 lines of code in a model definition you've made your model too complex for the use case (no product is that complicated).

One of the benefits is being able to sub-class, so you could create standard `Core`, `Life1`, `Life2`, `Expenses`, classes (say) and subclass from these (like mixins), and build a model from these, they might include:

Core
- month and year of projection (e.g. 5, 2023)
- valuation month
- is new year

Life1
- current age in years and months
- qx
- tpx

Life2
- as Life1
- joint distributions (first death & last survivor)

Expenses
- initial expenses
- maintenance expenses
- commissions
- total costs (sum of above)

Could then create a model class with just a few extra definitions?  (and override anything that doesn't fit the product)

```python
class Term(Model, Life1, Core, Expenses):
    def claim(self, t):
        return self.data['sum_assured'] * self.life1_qx_monthly(t)  # this is defined in Life1
```

Edit: added a code example here (not vectorised): https://github.com/lewisfogden/heavytables/blob/main/examples/hl_inherit.py

**Reply by MatthewCaseres** [reply]
Unsure how to feel about this

```
class Model ...

        for k, v in kwargs.items():
            if k in dir(self):
                warnings.warn("Warning: Duplicate Item: "+str(k))
            setattr(self, k, v)
```

The type checker does not like it

<img width="348" alt="Screenshot 2024-03-16 at 8 54 06 AM" src="https://github.com/actuarialopensource/actuarialopensource.org/assets/43053796/fee11464-f734-4f27-9723-25bcd6702ae3">

I don't think all things we pass in should become an attribute of the class, this is defining custom behavior that I feel is not pythonic.

Can we just do it like this?

```py
class Core(Model):

    def __init__(self, proj_len, do_run, data):
        self.data = data
        super().__init__(proj_len=proj_len, do_run=do_run)

    def t(self, t):
        return t
    
    def proj_month(self, t):
        """projection month: 1=Jan, 12=Dec"""
        if t == 0:
            return self.data['init_proj_month']
        elif self.proj_month(t - 1) == 12:
            return 1
        else:
            return self.proj_month(t - 1) + 1
```

This thread has some discussion - https://python-forum.io/thread-39422.html.

edit: I want to implement this as a decorator on the class and then inheritance can be avoided entirely. Not sure how I feel about mixins, it goes against the idea of "composition over inheritance". I would rather just dependency inject Life1 into the constructor of whatever needs it.

**Reply by lewisfogden** [reply]
Yeah I did more for the flexibility, particularly different models will have different inputs, e.g. I would see common being `data`, `basis`, `stresses` and `params` which might capture policy data, assumptions, and product features (e.g. maximum term), but you can also pass in items directly, and not neccesarily in a dictionary.

I'm focussing on the Table class at the moment, as that is useful for all python-based models (not just mine).  The look-up performance looks really good.   I've added an example notebook which is a start to documentation, and am thinking about the right approach for me for adding bounds (e.g. if mortality tables have a 5 year select period, then any duration above 5 should look up 5)

https://github.com/lewisfogden/heavytables/blob/main/examples/table_examples.ipynb

**Reply by lewisfogden** [reply]
Added the bounded code, which just pops a numpy.clip around input_data, some output:

<img width="602" alt="Screenshot 2024-03-17 at 14 10 31" src="https://github.com/actuarialopensource/actuarialopensource.org/assets/8605587/5240c9c7-e9b5-4092-8b68-4e13522d6a2f">

**Reply by MatthewCaseres** [reply]
I'm working on setting up CI/CD pipelines (pytest/codecov for now) + developer container config files so people can just jump into GitHub codespaces and do dev work. Not starting testing yet, just setting up the scaffolding. PR to come in the next few hours or so.

**Reply by MatthewCaseres** [reply]
pull request is out there. I'm going to be looking into implementing heavylight but just doing proof of concept with different approaches.

**Reply by MatthewCaseres** [reply]
After some thought it seems like heavylight handles things in a reasonable way using inheritance. So I guess I am pretty much on board with the major design decisions. 

edit: going through the heavylight code more, there is sort of a core of it and then there is some other details that could be implemented by the user. Things like `do_run`, `BeforeRun`, and `AfterRun` all feel like they might not need to be part of a library but can be implemented by the user. Ultimately the caching of the functions is so simple that it is reasonable for people to not rely on any packages and just do it themselves.

**Reply by lewisfogden** [reply]
Agree, I probably don't need `BeforeRun` and `AfterRun`.  `do_run` is almost always going to be true so could probably be defaulted, perhaps using proj_len=0 as the default that effectively doesn't precompute any of the outputs would remove the need to have `do_run` altogether.

Code is MIT for both so please play around - I'm thinking of keeping heavytables.Table as a separate package as I think it could be a very useful library in general.

I've not designed heavylight to be fast - it's is aimed to be easy (minimum boilerplate) to help wean people off Excel based-modelling (which still has a place). For example, using dictionaries rather than decorators is about twice as fast.

I've done a little work on a cython version (can't use decorators) and a few other versions - I've invited you to the private repo I have for it so you can have a look.

**Reply by MatthewCaseres** [reply]
We need a big benchmark. I want for the model to be running for like at least 10 seconds. Preferably more. Actuaries where I worked used to wait hours on a computing grid, I wonder what they were doing?

I like that heavymodels don't create this global cache state. I can make multiple instances of a model and compare them. The features I want are the following:
* Memory optimized runs
* Ability to unit test formulas
* Easy creation of results dataframe

I get that it is MIT license sure I can play around with it, but I really want to be using the same software as you and not just having some fork. Then we have two people doing things the same way, and maybe it can grow. Anyhow, I've been spending much time thinking and I don't know how productive it has been. I'm going to try to make a decent pull request on heavylight, the PR on heavytables is not so great but I thought if there was scaffolding for unit testing it would be useful.

**Reply by MatthewCaseres** [reply]
going offline until PR. might take a moment

**Reply by lewisfogden** [reply]
Thanks, I'm aiming to spend a week or so trialling heavytables out and checking it can cope with all my use cases.  I've got a Table library in heavylight at the moment but it's pandas series based and a bit iffy, so might merge the two (although heavytables is probably useful unbundled too).

Happy to join forces, I am already using heavylight (unfortunately don't know who else is) so I need to be careful about breaking changes (and put in warnings etc if they are).

Worth putting together a roadmap/idea list?

Alongside yours, I'd have
- pip/pypi install package
- user examples installable from package (included in pip)
- documentation / API written
- HOWTOs for new users
- on tests - excel models and heavylight equivalent models demonstrating same results - people trust Excel
- More of a package/suite approach - not a full blown UI
- Controlled way of running production jobs

**Reply by MatthewCaseres** [reply]
Here is the source code for the Class to inherit from https://github.com/actuarialopensource/heavylight/blob/main/src/heavylight/heavylight.py

You can see that the only public facing methods `RunModel` and `ToDataFrame` have not changed. Naming conventions for which methods are cached have not changed either.

Happy things first, scroll to the bottom to see a list of breaking changes, let me know if there are any deal breakers.

## Previous TODOs are done + implementation details

* Implemented code coverage and automated testing
  * 100% code coverage (Still need to test more things, like `t` comes first in `ToDataFrame` for example)
* All of the caches can be cleared and the model can be rerun
  * This is necessary for memory optimization which uses a warm-up run
* Implemented memory saving mechanism which is [showing over 99% reduction in cache size](https://github.com/actuarialopensource/heavylight/blob/a7f6b702528da0e9583cb514dcfbbba3f9ed90c9/tests/test_memory_optimizations.py#L62)
* There is now a decorator and a class to inherit from. The class iterates through its methods and applies the decorator. In theory you can just use the decorator though, but not sure if I will document this or just leave it as an implementation detail.
* There is an intermediate dictionary representation of stored results which is easy to load into a DataFrame or perform transformations on
  * Can easily aggregate/round/select from cached NumPy arrays before loading them into a DataFrame.
  
## Plans for the weekend

* Raise error if user runs twice without resetting the caches
* Implement and test a simpler non-optimizable cache to be used in situations where the user does not need additional performance
  * May be two classes, OptimizableModel and Model, using two different decorators.
* Build out the documentation site, may focus on documenting the non-optimizable cache first because it is simpler to explain.

I'm going to have to give an extra effort at work and school starting Monday so the plan is to have a PR on Sunday night and then you can think on it and give feedback. Unsure how far I will get.

The points on your roadmap are noted and I have no disagreements, but can't get them done by Sunday.

## Notes on testing of models

### Providing Excel reference implementations

I agree that reference implementations in Excel are good if we are building out some standard models `lifelib` style. 

### Grinding out coverage to 100%

My view of testing is really about being able to write tests in PyTest. Basically I am going to define `__setitem__` on your `_Cache`'d method. And then I am going to write a unit test like

```
model.pols_if[1] = 1
model.pols_death[1] = .5
assert model.pols_if(2) == .5
```

This is just because there is a lot of interconnectedness in the methods and being able to inject values into the cache will allow us to isolate this. Specifically I am trying to ensure we are able to grind out code coverage to 100% on any models with unit tests. Obviously we can integration test, but as soon as we swap out a mortality table, the integration tests break. **To unit test, we inject values in the cache.**

This is implemented in fork - https://github.com/actuarialopensource/heavylight/blob/a7f6b702528da0e9583cb514dcfbbba3f9ed90c9/src/heavylight/cache_graph.py#L102

## Breaking changes

* Removal of tables class
  * This code is not related to the caching of methods
* Removal of `verbose`
  * Seeing a lot more of this than I am used to seeing in other open source projects. I think you are using it to debug while developing but maybe using the IDE's debugger will be a suitable replacement?
* Not dynamically assigning object properties based on what the user passed in
  * The user will super init
* methods must have a single parameter **and it must be named t** for them to be called on `RunModel`.
  * I know you are not requiring `t` to be the param name, but if we are going to be iterating over timesteps and calling this method, I think it is reasonable to ask for

**Reply by lewisfogden** [reply]
Thank you - will have a look over.  Main comment is that single parameter methods won't work for some of my production use cases, so maybe that needs to be optional (i.e. I've set it so if one parameter then it will loop it, maybe the constraint is it only loops it if it is labelled `t`?)

I'm tempted to say we should treat this as a community driven experimental model (or multiple models) and not necessarily heavylight v2.  (Aim of heavylight is to give a simple way to generate projections, which IT security departments can check easily).

I've been working on testing of heavytables which I've merged into heavylight (and get rid of the original tables class as it wasn't great) - maybe we should keep these separate though?  (Found one problem bug, with false positives on string lookup)

There are a lot of good ideas in cashflower too (https://github.com/acturtle/cashflower/tree/main by @zchmielewska et al.).

**Reply by MatthewCaseres** [reply]
> I've set it so if one parameter then it will loop it, maybe the constraint is it only loops it if it is labelled t?

That is how it currently works in my codebase. Sorry if unclear.

> I'm tempted to say we should treat this as a community driven experimental model (or multiple models) and not necessarily heavylight v2. 

In this case I guess I would take my stuff, put it in it's own `memory_optimized.py` file, and then send that over?

> Aim of heavylight is to give a simple way to generate projections, which IT security departments can check easily

Unsure what security concerns I've introduced?

> I've been working on testing of heavytables which I've merged into heavylight (and get rid of the original tables class as it wasn't great) - maybe we should keep these separate though? (Found one problem bug, with false positives on string lookup)

Can you set up some automated testing? Or I guess if my PR goes through then you will have it. It usually catches a couple bugs and makes you feel safe when refactoring code.

## my main takeaway

I was wanting to rewrite some of the existing code because of things that bother me like `do_run` or dynamic assignment of object properties. This is probably not needed as there isn't really anything wrong with it, it just doesn't follow a certain style that I like, and that is fine. **no attempts to delete existing source code, additive changes only**.

My main complaint about cashflower is that it is opinionated, describing should files be named, how should assumptions be accessed. But I do like their aggregation type. I implement it by letting a user define a function.

https://cashflower.acturtle.com/model.html#aggregation-type

**Reply by lewisfogden** [reply]
I'm a bit of a noob at writing testing (and github in general), and it's something I need to work on - I think I approved your PR already?

My main concern is you spend lots of time writing code and I have to say no to it, so probably just need to do a bit of planning.  Like the idea of getting rid of `do_run` and maybe just defaulting proj_len to 0 to achieve the same thing.

Security is just about size of code - anything I bring into my org needs to be scanned and have due dilligence, so less code = easier to check for issues.  Obviously heavylight is a bit easier as I've written it, but modelx/lifelib/cashflower are all big libraries with a lot of code I'd need to check.

I've got a branch with the integrated Tables in, am working on documentation (and found a flaw in my string lookup approach `In [10`] here):
https://github.com/lewisfogden/heavylight/blob/dev_heavytable_integration/examples/notebook/table_documentation.ipynb

**Comment by lewisfogden** [comment]
This chain is getting a little hard to follow from all the ideas.  @MatthewCaseres - I've added you to the repo (and set up a project) - does that work for you?  Let's get issues/ideas noted out as separate items.

**Reply by MatthewCaseres** [reply]
Sure.
